File: site_pages/schema.py
"""Open schema generator: JSON Resume + Schema.org Person.

Produces two machine-readable representations of the portfolio, both derived
from the same inputs -- the parsed identity config (``site.yml``) and the
content records loaded from ``content/`` by :mod:`site_pages.loader`:

1. A **JSON Resume** document (https://jsonresume.org/spec/) served at
   ``/.well-known/portfolio.json``. Spec-strict for ``basics`` and ``projects``;
   blog posts and tutorials have no native JSON Resume home, so they're exposed
   under a clearly-namespaced ``_extensions`` block (leading underscore marks it
   as non-spec; spec-strict consumers ignore it).
2. A **Schema.org ``Person``** JSON-LD blob, embedded inline in the page
   ``<head>`` for crawlers.

Both generators are pure functions of their inputs, which makes them trivial to
test and reuse.
"""
from __future__ import annotations

from typing import Any

from . import loader
from .context_processors import load_site_config

# Map site.yml link keys (and the networks agents care about) to JSON Resume
# ``basics.profiles[].network`` labels. Any link not in the map is kept with a
# title-cased key so nothing is silently dropped.
_NETWORK_LABELS = {
    "github": "GitHub",
    "linkedin": "LinkedIn",
    "twitter": "Twitter",
    "mastodon": "Mastodon",
    "website": "Website",
    "youtube": "YouTube",
}


def _network_label(key: str) -> str:
    return _NETWORK_LABELS.get(key, key.replace("_", " ").title())


def build_json_resume(site: dict | None = None) -> dict[str, Any]:
    """Build a JSON Resume document from site identity + content records.

    ``site`` defaults to the parsed ``site.yml``. Pass an explicit dict in tests
    to avoid touching the filesystem.
    """
    site = site if site is not None else load_site_config()
    links = site.get("links") or {}

    basics: dict[str, Any] = {
        "name": site.get("name") or "",
        "label": site.get("label") or "",
    }
    if site.get("email"):
        basics["email"] = site["email"]
    if site.get("base_url"):
        basics["url"] = site["base_url"]
    if site.get("location"):
        basics["location"] = {"address": site["location"]}
    profiles = [{"network": _network_label(k), "url": v} for k, v in links.items() if v]
    if profiles:
        basics["profiles"] = profiles

    # Projects map cleanly onto the JSON Resume "projects" section.
    projects = [_project_to_resume(p) for p in loader.load_collection("projects")]
    doc: dict[str, Any] = {"basics": basics}
    if projects:
        doc["projects"] = projects

    # Blog + tutorials have no JSON Resume home; expose under a namespaced
    # extension so the doc stays spec-strict but the writing is still readable.
    posts = _summaries(loader.load_collection("blog"))
    tutorials = _summaries(loader.load_collection("tutorials"))
    if posts or tutorials:
        doc["_extensions"] = {
            "schema": "https://portfolio-template.dev/extensions/v1",
            "posts": posts,
            "tutorials": tutorials,
        }
    return doc


def _project_to_resume(p: dict) -> dict[str, Any]:
    """Map one portfolio project record to a JSON Resume ``projects`` entry."""
    entry: dict[str, Any] = {
        "name": p.get("title") or p["slug"],
        "summary": p.get("summary") or "",
        "slug": p["slug"],
    }
    if p.get("tech"):
        entry["keywords"] = list(p["tech"])
    links = p.get("links") or {}
    if links.get("github"):
        entry["repo"] = links["github"]
    if links.get("live"):
        entry["url"] = links["live"]
    if p.get("date"):
        entry["startDate"] = p["date"].isoformat()
    return entry


def _summaries(records: list[dict]) -> list[dict[str, Any]]:
    """Reduce a collection to the lightweight fields useful for indexing."""
    out = []
    for r in records:
        entry: dict[str, Any] = {
            "slug": r["slug"],
            "title": r.get("title") or r["slug"],
            "summary": r.get("summary") or "",
        }
        if r.get("date"):
            entry["date"] = r["date"].isoformat()
        if r.get("tags"):
            entry["tags"] = list(r["tags"])
        out.append(entry)
    return out


def build_person_jsonld(site: dict | None = None) -> dict[str, Any]:
    """Build a Schema.org ``Person`` JSON-LD object for crawlers."""
    site = site if site is not None else load_site_config()
    person: dict[str, Any] = {
        "@context": "https://schema.org",
        "@type": "Person",
        "name": site.get("name") or "",
    }
    if site.get("base_url"):
        person["url"] = site["base_url"]
    if site.get("email"):
        person["email"] = site["email"]
    same_as = [v for v in (site.get("links") or {}).values() if v]
    if same_as:
        person["sameAs"] = same_as

    # knowsAbout = union of all project tech + post/tutorial tags.
    knows: set[str] = set()
    for p in loader.load_collection("projects"):
        knows.update(p.get("tech") or [])
    for kind in ("blog", "tutorials"):
        for r in loader.load_collection(kind):
            knows.update(r.get("tags") or [])
    if knows:
        person["knowsAbout"] = sorted(knows)
    return person

File: src/lib/schema.ts
// JSON Resume + Schema.org Person builders. Port of site_pages/schema.py.
// Pure functions of (site identity, content collections).

import type { SiteConfig } from "../config";
import { getCollection } from "astro:content";

const NETWORK_LABELS: Record<string, string> = {
  github: "GitHub",
  linkedin: "LinkedIn",
  twitter: "Twitter",
  mastodon: "Mastodon",
  website: "Website",
  youtube: "YouTube",
};

function networkLabel(key: string): string {
  return NETWORK_LABELS[key] || key.replace(/_/g, " ").replace(/\b\w/g, (c) => c.toUpperCase());
}

interface CollectionEntryLike {
  id: string;
  data: Record<string, unknown>;
}

function isDraft(entry: { data: { draft?: boolean } }): boolean {
  return entry.data.draft === true;
}

function summaryFields(entries: CollectionEntryLike[]): Record<string, unknown>[] {
  return entries.map((e) => {
    const d = e.data;
    const out: Record<string, unknown> = {
      slug: e.id,
      title: d.title || e.id,
      summary: d.summary || "",
    };
    if (d.date) out.date = new Date(d.date as string).toISOString().slice(0, 10);
    if (Array.isArray(d.tags) && d.tags.length) out.tags = d.tags;
    return out;
  });
}

async function loadAll(name: "projects" | "tutorials" | "blog"): Promise<CollectionEntryLike[]> {
  // getCollection is async in Astro v7 content layer.
  const showDrafts = import.meta.env.SHOW_DRAFTS === "true" || import.meta.env.DEV;
  const all = (await getCollection(name)) as unknown as CollectionEntryLike[];
  return showDrafts ? all : all.filter((e) => !isDraft(e as unknown as { data: { draft?: boolean } }));
}

export async function buildJsonResume(site: SiteConfig): Promise<Record<string, unknown>> {
  const basics: Record<string, unknown> = {
    name: site.name || "",
    label: site.label || "",
  };
  if (site.email) basics.email = site.email;
  if (site.base_url) basics.url = site.base_url;
  if (site.location) basics.location = { address: site.location };
  const profiles = Object.entries(site.links)
    .filter(([, v]) => v)
    .map(([k, v]) => ({ network: networkLabel(k), url: v }));
  if (profiles.length) basics.profiles = profiles;

  const projects = (await loadAll("projects"))
    .sort((a, b) => {
      const da = a.data.date ? new Date(a.data.date as string).getTime() : 0;
      const db = b.data.date ? new Date(b.data.date as string).getTime() : 0;
      return db - da;
    })
    .map((p) => {
      const d = p.data;
      const links = (d.links || {}) as { github?: string; live?: string };
      const entry: Record<string, unknown> = {
        name: d.title || p.id,
        summary: d.summary || "",
        slug: p.id,
      };
      if (Array.isArray(d.tech) && d.tech.length) entry.keywords = d.tech;
      if (links.github) entry.repo = links.github;
      if (links.live) entry.url = links.live;
      if (d.date) entry.startDate = new Date(d.date as string).toISOString().slice(0, 10);
      return entry;
    });

  const doc: Record<string, unknown> = { basics };
  if (projects.length) doc.projects = projects;

  const posts = summaryFields(await loadAll("blog"));
  const tutorials = summaryFields(await loadAll("tutorials"));
  if (posts.length || tutorials.length) {
    doc._extensions = {
      schema: "https://portfolio-template.dev/extensions/v1",
      posts,
      tutorials,
    };
  }
  return doc;
}

export async function personJsonLd(site: SiteConfig): Promise<Record<string, unknown>> {
  const person: Record<string, unknown> = {
    "@context": "https://schema.org",
    "@type": "Person",
    name: site.name || "",
  };
  if (site.base_url) person.url = site.base_url;
  if (site.email) person.email = site.email;
  const sameAs = Object.values(site.links).filter(Boolean);
  if (sameAs.length) person.sameAs = sameAs;

  const knows = new Set<string>();
  for (const p of await loadAll("projects")) {
    if (Array.isArray(p.data.tech)) p.data.tech.forEach((t: string) => knows.add(t));
  }
  for (const kind of ["blog", "tutorials"] as const) {
    for (const r of await loadAll(kind)) {
      if (Array.isArray(r.data.tags)) r.data.tags.forEach((t: string) => knows.add(t));
    }
  }
  if (knows.size) person.knowsAbout = Array.from(knows).sort();
  return person;
}
